File: FormalBench-data/FilteredData/specifications/seed_726_per_category_10_653ade686f/c/LeftInsertion.c
#include "java_arrays.h"
#include <stdint.h>

/*@
assigns \nothing;
ensures \result == (int32_t)((integer)left + right);
*/
static int32_t java_add(int32_t left, int32_t right) {
    return (int32_t)((uint32_t)left + (uint32_t)right);
}

/*@
assigns \nothing;
ensures \result == (int32_t)((integer)left - right);
*/
static int32_t java_sub(int32_t left, int32_t right) {
    return (int32_t)((uint32_t)left - (uint32_t)right);
}

/*@
requires divisor != 0;
assigns \nothing;
ensures \result == (dividend == INT32_MIN && divisor == -1 ? INT32_MIN : (integer)dividend / divisor);
*/
static int32_t java_div(int32_t dividend, int32_t divisor) {
    if (dividend == INT32_MIN && divisor == -1) {
        return INT32_MIN;
    }
    return dividend / divisor;
}

/*@
logic integer fb_mid(integer lo, integer hi) = (int32_t)(lo + hi) / 2;
logic integer fb_search{L}(JIntArray a, integer x, integer lo, integer hi) =
  lo > hi ? lo : fb_mid(lo, hi) < 0 || fb_mid(lo, hi) >= a->length ? -1 :
  a->data[fb_mid(lo, hi)] == x ? fb_mid(lo, hi) : a->data[fb_mid(lo, hi)] < x ?
  fb_search(a, x, fb_mid(lo, hi) + 1, hi) : fb_search(a, x, lo, fb_mid(lo, hi) - 1);
*/
/*@
requires jintarray_valid_nonnull(a);
requires fb_search(a, x, 0, a->length - 1) >= 0;
assigns \nothing;
ensures \result == fb_search(a, x, 0, a->length - 1);
ensures 0 <= \result <= a->length;
ensures (\forall integer p, q; 0 <= p < q < a->length ==> a->data[p] <= a->data[q]) ==>
  (\forall integer k; 0 <= k < \result ==> a->data[k] <= x) &&
  (\forall integer k; \result <= k < a->length ==> x <= a->data[k]);
*/
int32_t leftInsertion(JIntArray a, int32_t x) {
    int32_t left = 0;
    int32_t right = java_sub(jarray_length(a), 1);
    /*@
    loop invariant 0 <= left <= a->length && -1 <= right < a->length && left <= (integer)right + 1;
    loop invariant fb_search(a, x, 0, a->length - 1) == fb_search(a, x, left, right) && fb_search(a, x, left, right) >= 0;
    loop assigns left, right;
    loop variant (integer)right - left + 1;
    */
    while (left <= right) {
        int32_t mid = java_div(java_add(left, right), 2);
        int32_t value = jarray_get(a, mid);
        if (value == x) {
            return mid;
        } else if (value < x) {
            left = java_add(mid, 1);
        } else {
            right = java_sub(mid, 1);
        }
    }
    return left;
}

File: FormalBench-data/FilteredData/specifications/seed_726_per_category_10_653ade686f/java/LeftInsertion.java

import java.io.*;
import java.lang.*;
import java.util.*;
import java.math.*;

class LeftInsertion {
    
    /*@
    public normal_behavior
      ensures \result == ((x % 4294967296L + 6442450944L) % 4294967296L) - 2147483648L;
    model public static pure \bigint fb_wrap(\bigint x);

        public normal_behavior
          ensures \result == fb_wrap((\bigint)lo + hi) / 2;
        model public static pure int fb_mid(int lo, int hi);
        public normal_behavior
          requires a != null && 0 <= lo <= a.length && -1 <= hi < a.length && lo <= (\bigint)hi + 1;
          ensures \result == (lo > hi ? lo : fb_mid(lo, hi) < 0 || fb_mid(lo, hi) >= a.length ? -1 :
            a[fb_mid(lo, hi)] == x ? fb_mid(lo, hi) : a[fb_mid(lo, hi)] < x ?
            fb_search(a, x, fb_mid(lo, hi) + 1, hi) : fb_search(a, x, lo, fb_mid(lo, hi) - 1));
          measured_by (\bigint)hi - lo + 1;
        model public static pure int fb_search(int[] a, int x, int lo, int hi);

        public normal_behavior
          requires a != null && fb_search(a, x, 0, a.length - 1) >= 0;
          assignable \nothing;
          ensures \result == fb_search(a, x, 0, a.length - 1);
          ensures 0 <= \result <= a.length;
          ensures (\forall int p, q; 0 <= p < q < a.length; a[p] <= a[q]) ==>
            (\forall int k; 0 <= k < \result; a[k] <= x) && (\forall int k; \result <= k < a.length; x <= a[k]);
          also
        public exceptional_behavior
          requires a == null;
          assignable \nothing;
          signals_only NullPointerException;
          also
        public exceptional_behavior
          requires a != null && fb_search(a, x, 0, a.length - 1) == -1;
          assignable \nothing;
          signals_only ArrayIndexOutOfBoundsException;
    @*/
    //@ code_java_math
    public static int leftInsertion(int[] a, int x) {
        int left = 0;
        int right = a.length - 1;
        /*@
        loop_invariant 0 <= left <= a.length && -1 <= right < a.length && left <= (\bigint)right + 1;
        loop_invariant fb_search(a, x, 0, a.length - 1) == fb_search(a, x, left, right);
        loop_writes left, right;
        decreases (\bigint)right - left + 1;
        @*/
        while (left <= right) {
            int mid = (left + right) / 2;
            if (a[mid] == x) {
                return mid;
            } else if (a[mid] < x) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return left;
    }
}

